src/meta/src/hummock/compaction/level_selector.rs

```diff
@@ -82,6 +88,81 @@ impl DynamicLevelSelector {
             inner: LevelSelectorCore::new(config, overlap_strategy),
         }
     }
+
+    fn calculate_l0_overlap(
+        &self,
+        l0: &OverlappingLevel,
+        base_level: &Level,
+        handlers: &[LevelHandler],
+    ) -> u64 {
+        let total_level_size = l0.total_file_size - handlers[0].get_pending_file_size();
+        let mut overlap_info = self.inner.overlap_strategy.create_overlap_info();
+        for sub_level in &l0.sub_levels {
+            for table_info in &sub_level.table_infos {
+                if handlers[0].is_pending_compact(&table_info.id) {
+                    overlap_info.update(table_info);
+                }
+            }
+        }
+        let mut next_level_size = 0;
+        let output_files = handlers[base_level.level_idx as usize].get_pending_output_file();
+        let overlap_files = overlap_info.check_multiple_overlap(&base_level.table_infos);
+
+        for sst in &overlap_files {
+            // this file would not stay in current-level because this data would be moved to the
+            // next level. But for other files, even if it is pending in another compact
+            // task and would be deleted after compact task end, the data of which would
+            // still stay in this level.
+            if output_files.contains(&sst.id) {
+                continue;
+            }
+            next_level_size += sst.file_size;
+        }
+        next_level_size + total_level_size
+    }
+
+    fn calculate_base_level_overlap(
+        &self,
+        target_bytes: u64,
+        select_level: &Level,
+        target_level: &Level,
+        handlers: &[LevelHandler],
+    ) -> u64 {
+        if select_level.total_file_size <= target_bytes {
+            return 0;
+        }
+        let compacting_file_size = handlers[select_level.level_idx as usize]
+            .get_pending_output_file_size(target_level.level_idx);
+        if select_level.total_file_size - compacting_file_size <= target_bytes {
+            return 0;
+        }
+        let output_files = handlers[select_level.level_idx as usize].get_pending_output_file();
+        let mut info = self.inner.overlap_strategy.create_overlap_info();
+        let mut compact_bytes = 0;
+        for sst in &select_level.table_infos {
+            if compact_bytes + compacting_file_size + target_bytes >= select_level.total_file_size {
+                break;
+            }
+            if output_files.contains(&sst.id) {
```

Comment:
Why do we ignore pending SSTs output to select level?

Reply:
fixed